oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -0,0 +1,168 @@
+                        String reindexCompletionTime = indexDefParentNode
+                                .getChildNode(indexNameObjectList.get(i).getNodeName())
+                                .getChildNode(IndexDefinition.STATUS_NODE)
+                                .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP).getValue(Type.DATE);
+                        long reindexCompletionTimeInMillis = PurgeOldVersionUtils.getMillisFromString(reindexCompletionTime);
+                        long currentTimeInMillis = System.currentTimeMillis();
+                        if (currentTimeInMillis - reindexCompletionTimeInMillis > threshold) {
+                            activeProductVersion = indexNameObjectList.get(i).getProductVersion();
+                        }
+                    } else {
+                        log.warn(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP
+                                + " property is not set for index " + indexNameObjectList.get(i).getNodeName());
+                    }
+                }
+            } else {
+                if (indexNameObjectList.get(i).getProductVersion() == activeProductVersion
+                        && indexNameObjectList.get(i).getCustomerVersion() == 0) {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE_HIDDEN_AND_DISABLE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                } else {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                }
+            }
+        }
+        if (!isValidIndexVersionOperationList(indexVersionOperationList)) {
+            indexVersionOperationList = Collections.EMPTY_LIST;
+        }
+        return indexVersionOperationList;
+    }
+
+    /*
+        returns IndexNameObjects in descending order of version i.e from latest to oldest
+    */
+    private static List<IndexName> getSortedVersionedIndexNameObjects(List<IndexName> indexNameObjectList) {
+        indexNameObjectList.sort(new Comparator<IndexName>() {
+            @Override
+            public int compare(IndexName indexNameObj1, IndexName indexNameObj2) {
+                return indexNameObj2.compareTo(indexNameObj1);
+            }
+        });
+        return indexNameObjectList;
+    }
+
+    private static boolean isValidIndexVersionOperationList(List<IndexVersionOperation> indexVersionOperations) {
+        boolean isValid = false;
+        IndexVersionOperation lastNoopOperationIndexVersion = null;
+        IndexVersionOperation indexWithDeleteHiddenOp = null;
+        for (int i = 0; i < indexVersionOperations.size(); i++) {
```

Comment:
I would use the enhanced for and avoid the number of calls to `indexVersionOperations.get(i)`